[wip] Use pure zsh to load bundles

forivall opened the pull request:
The original implementation runs way more processes than are necessary.
While it is normally good practice to use standard unix conventions in
order to maintain compatability, commands that run in a zshrc file
should ideally be as fast as possible. Using sed, awk, tr, etc.
drastically slow down shell initialization time, _especially_ on cygwin,
because processes take much longer than on a native unix platform.

This may alleviate the need for #129.

* Should the other commands be updated to not create extra processes / use pure zsh?
* How much should i try to use fancy zsh flags? or should i stick with `regexp-replace` and `-antigen-arraygrep`
* I need help getting the tests to run, the basic `pip install` + `make` didn't work on cygwin (i'll try my linux install later)
* should i just switch to using zgen?
* what old versions of zsh should be supported? how much do the tests cover? (hopefully travis does this, too tired rn to look)

desyncr commented:
While I do agree that 'grep', 'sed' and so on should be used scarcely it seems it doesn't improve performance on other than cygwin environments, at least here (Fedora 22, zsh 5.0.7) the start up time is as follow:

```
# performance-improvements branch
$ time zsh
zsh  0.35s user 0.17s system 104% cpu 0.499 total

# master
time zsh
zsh  0.38s user 0.44s system 113% cpu 0.718 total
```

While the [cache branch](https://github.com/zsh-users/antigen/pull/129) would improve start up time through out all platforms and environments.


> This may alleviate the need for #128.

I don't think antigen completion as anything to do with performance.

> Should the other commands be updated to not create extra processes / use pure zsh?

It should be applied uniformly through out all antigen code if possible as to not have too many ways to do one thing in the code.

> How much should i try to use fancy zsh flags?

As long as it doesn't introduce backward incompatibility.

forivall commented:
Whoops, that was supposed to be
> This may alleviate the need for #129.

Yeah, it doesn't have anything to do with completion.

desyncr commented:
This doesn't alleviate the need for cache as the performance improvements it brings turn start up time from >0.40s to 0.10s and aren't platform specific.

forivall commented:
I didn't say that it would alleviate the need for cache, I was just pondering if it would. Also, the cache branch isn't ready, and requires user intervention to get it to work (currently), will require cache invalidation, etc.. This is a much smaller, simpler, less complex change, and it does provide a significant performance increase, especially when lots of bundles are used.

To compare, here is the difference in time provided by caching on my system:
```
# build cache
$ time zsh
zsh  0.67s user 0.14s system 44% cpu 1.822 total

# load cache
$ time zsh
zsh  0.63s user 0.08s system 79% cpu 0.886 total
```

So, you can see, even on linux, my patch does lessen the need for caching; caching only provides a 15% speed increase over my `performance-improvements`.

desyncr commented:
@forivall I looked at your configs (thanks for pointing them out). Given your configs nothing is being cached as 'antigen bundle' is used instead of 'antigen bundles'.

Changing the configuration would give the following stats:

```
$ time zsh
zsh  0.66s user 0.92s system 119% cpu 1.326 total

$ time zsh
zsh  0.38s user 0.09s system 102% cpu 0.461 total
```

forivall commented:
@desyncr I hadn't pushed my changes yet. It does use the cache. Check again.

desyncr commented:
@forivall I'm sorry -zcache-start and -zcache-done should only be called inside antigen, not from outside. I don't know how it would behave when using it in a non-standard way.

forivall commented:
It behaves exactly the same. I [changed](https://github.com/forivall/dotfiles/commit/c36b3238c47c8ad85660105f60e62ad7beda6ea0) [my dotfiles](https://github.com/forivall/dotfiles) to use `antigen bundles`, and the performance is exactly the same as shown above:
```
# performance-improvements*
$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.67s user 0.08s system 77% cpu 0.966 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.70s user 0.11s system 80% cpu 1.009 total

$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.


# master*
$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.64s user 0.18s system 45% cpu 1.800 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.69s user 0.15s system 46% cpu 1.825 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.68s user 0.13s system 45% cpu 1.795 total

$ git checkout feature/bundle-cache-patch-1 
Switched to branch 'feature/bundle-cache-patch-1'


# feature/bundle-cache-patch-1
$ time PERF_TEST=y zsh                     
PERF_TEST=y zsh  0.61s user 0.09s system 81% cpu 0.865 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.62s user 0.08s system 79% cpu 0.880 total

$ rm .cache/.*

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.69s user 0.15s system 44% cpu 1.903 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.67s user 0.08s system 81% cpu 0.917 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.71s user 0.09s system 82% cpu 0.973 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.65s user 0.11s system 80% cpu 0.936 total

$ time PERF_TEST=y zsh
PERF_TEST=y zsh  0.69s user 0.10s system 82% cpu 0.953 total
```
The other half of the speed bonus is that you're running your dotfiles are stored on an ssd. I have my dotfiles stored on a [hybrid hdd](http://www.newegg.com/Product/Product.aspx?Item=N82E16822178379).

forivall commented:
Screw it, I'm switching to zgen. It's just as fast as my tweaks, and I don't have to deal with this memester.

forivall commented:
If somebody wants to take this and clean it up, that's fine by me, but any further involvement from me will be minimal.

forivall commented:
@HeroCC You can also try out this branch; it makes safer changes than the cache branch for similar performance gains.

blairham commented:
I've been using this branch for sometime now, no problems can we get this merged in?

paulhybryant commented:
+1 for merging this.

blairham commented:
This looks to be a dead project, time to clone and branch it into a new one.

forivall commented:
@blairham I would suggest looking at zgen, zplugin or antibody

blairham commented:
Thanks @forivall ;) I feel you're frustration on this project. it's been working fine but the lack of activity on it bothers me. I hadn't switched yet as I really like how antigen works.

desyncr commented:
For the reason exposed above and the lack of interest in maintaining/cleaning this PR I'm closing this out.

Thanks for using and contributing to Antigen.
